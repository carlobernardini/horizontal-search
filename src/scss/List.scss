.orizzonte__list {
    @extend %list;
    margin: 5px 0 0;
    min-width: $list-min-width;
    animation: listIn .1s linear;
    transform-origin: $list-transform-origin;

    &--right {
        left: auto;
        right: 0;
        transform-origin: top right;
    }

    &:hover {
        .orizzonte__list-controls--mouseover {
            margin: 5px 10px;
            transform: scale(1);
            max-height: 100px;
            opacity: 1;
            transition-delay: 0s;
        }
    }
}

.orizzonte__list-controls {
    display: flex;
    margin: 0 10px 5px;

    &--mouseover {
        max-height: 0;
        opacity: 0;
        transform: scale(.9);
        overflow: hidden;
        transition: all .3s;
        transition-delay: .5s;
    }
}

.orizzonte__list-control {
    outline: none;
    border: none;
    color: #fff;
    padding: 5px 0;
    cursor: pointer;
    transition: all .3s;
    display: table-cell;
    text-align: center;
    flex: 1;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    @extend %fonts;

    + .orizzonte__list-control {
        margin-left: 3px;
    }
}

.orizzonte__list-done {
    background-color: $btn-done-color;
    border-radius: 5px;
    flex-grow: 1;
    font: {
        weight: 500;
    }

    &:hover {
        background-color: lighten($btn-done-color, 5%);
    }
}

.orizzonte__list-remove {
    background-color: #FFF;
    color: $btn-remove-color;
    transition: color .3s;
}

.orizzonte__item {
    min-width: 175px;

    &--filters {
        + .orizzonte__item--filters {
            padding-top: 5px;
            margin-top: 5px;
            border-top: 1px solid #eee;
        }
    }
}

.orizzonte__item-clickable {
    @extend %list-item;
}
