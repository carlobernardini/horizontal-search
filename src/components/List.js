import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import '../scss/List.scss';

class List extends Component {
    renderDoneBtn() {
        const {
            doneBtn, doneBtnLabel, isFilterGroup, onApply
        } = this.props;

        if (!isFilterGroup || !doneBtn) {
            return null;
        }

        return (
            <li>
                <button
                    type="button"
                    className="orizzonte__list-done"
                    onClick={ onApply }
                >
                    { doneBtnLabel || 'Done' }
                </button>
            </li>
        );
    }

    renderItems() {
        const {
            values, items, isFilterGroup, onUpdate
        } = this.props;

        if (isFilterGroup) {
            return React.Children.map(items, (item, i) => (
                <li
                    className={ classNames('orizzonte__item', {
                        'orizzonte__item--filters': isFilterGroup
                    }) }
                    key={ i }
                >
                    { React.cloneElement(item, {
                        value: ((v, fn) => {
                            if (!(fn in v)) {
                                return null;
                            }
                            return v[fn];
                        })(values, item.props.fieldName),
                        onUpdate: (filterValue) => {
                            const { fieldName } = item.props;
                            onUpdate(fieldName, filterValue);
                        }
                    }) }
                </li>
            ));
        }

        return items.map((item, i) => (
            <li
                className={ classNames('orizzonte__item', {
                    'orizzonte__item--filters': isFilterGroup
                }) }
                key={ i }
            >
                { item }
            </li>
        ));
    }

    render() {
        const { orientation } = this.props;

        return (
            <ul
                className={ classNames('orizzonte__list', {
                    'orizzonte__list--right': orientation === 'right'
                }) }
            >
                { this.renderItems() }
                { this.renderDoneBtn() }
            </ul>
        );
    }
}

List.propTypes = {
    doneBtn: PropTypes.bool,
    doneBtnLabel: PropTypes.string,
    values: PropTypes.object,
    isFilterGroup: PropTypes.bool,
    items: PropTypes.array.isRequired,
    onApply: PropTypes.func,
    onUpdate: PropTypes.func,
    orientation: PropTypes.oneOf([
        'left',
        'right'
    ])
};

List.defaultProps = {
    doneBtn: true,
    doneBtnLabel: null,
    values: {},
    isFilterGroup: false,
    onApply: () => {},
    onUpdate: () => {},
    orientation: 'left'
};

export default List;
