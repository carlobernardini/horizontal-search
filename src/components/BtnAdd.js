import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { DISPLAY_NAME_BTN_ADD } from '../constants';
import List from './List';
import '../scss/Button.scss';

class BtnAdd extends PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            active: false
        };
        this.btnAdd = React.createRef();
        this.toggleButton = this.toggleButton.bind(this);
        document.addEventListener('click', this.onBodyClick.bind(this), false);
    }

    componentWillUnmount() {
        document.removeEventListener('click', this.onBodyClick.bind(this), false);
    }

    onBodyClick(e) {
        if (!this.btnAdd || !this.btnAdd.current || this.btnAdd.current.contains(e.target)) {
            return false;
        }

        this.setState({
            active: false
        });

        return true;
    }

    toggleButton() {
        const { active } = this.state;

        this.setState({
            active: !active
        });
    }

    renderList() {
        const { active } = this.state;
        const { available, onGroupAdd, position, shown } = this.props;

        if (!shown || !active) {
            return null;
        }

        return (
            <List
                orientation={ position === 'right' ? 'left' : 'right' }
            >
                { available.map((filter, i) => (
                    // eslint-disable-next-line jsx-a11y/anchor-is-valid
                    <a
                        key={ `orizzonte-add-${ i }` }
                        href="#"
                        className="orizzonte__item-clickable"
                        onClick={ (e) => {
                            e.preventDefault();
                            onGroupAdd(filter.i, filter.label);
                            this.toggleButton();
                        }}
                    >
                        { filter.label }
                    </a>
                )) }
            </List>
        );
    }

    render() { 
        const {
            available, disabled, label, position, shown
        } = this.props;

        return (
            <div
                className={ classNames('orizzonte__btn-wrapper', {
                    'orizzonte__btn-wrapper--left': position === 'left'
                }) }
                ref={ this.btnAdd }
            >
                <button
                    className={ classNames('orizzonte__btn orizzonte__btn-add', {
                        'orizzonte__btn--shown': shown,
                        'orizzonte__btn--disabled': disabled || !available.length,
                        'orizzonte__btn--labeled': label && label.trim().length
                    }) }
                    disabled={ !available.length }
                    onClick={ this.toggleButton }
                    type="button"
                >
                    { label || String.fromCharCode(160) }
                </button>
                { this.renderList() }
            </div>
        );
    }
}

BtnAdd.displayName = DISPLAY_NAME_BTN_ADD;

BtnAdd.propTypes = {
    available: PropTypes.array,
    disabled: PropTypes.bool,
    label: PropTypes.string,
    onGroupAdd: PropTypes.func,
    position: PropTypes.string,
    shown: PropTypes.bool
};

BtnAdd.defaultProps = {
    available: [],
    disabled: false,
    label: null,
    onGroupAdd: () => {},
    position: null,
    shown: false
};

export default BtnAdd;
