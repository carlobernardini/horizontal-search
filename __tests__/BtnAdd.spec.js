import React from 'react';
import BtnAdd from '../src/components/BtnAdd';
import List from '../src/components/List';

const onGroupAdd = jest.fn();

describe('<BtnAdd />', () => {
    it('should render a default add-button', () => {
        const wrapper = shallow(
            <BtnAdd
                available={ [{
                    label: 'Filter 1'
                }, {
                    label: 'Filter 2'
                }] }
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render an add-button that is (always) shown', () => {
        const wrapper = shallow(
            <BtnAdd
                available={ [{
                    i: 0,
                    label: 'Filter 1'
                }, {
                    i: 1,
                    label: 'Filter 2'
                }] }
                onGroupAdd={ onGroupAdd }
                shown
            />
        );

        expect(wrapper).toMatchSnapshot();
        wrapper.find('.orizzonte__btn-add').simulate('click');
        expect(wrapper.state().active).toBe(true);
        wrapper.find(List).childAt(0).simulate('click', {
            preventDefault: () => {}
        });
        expect(onGroupAdd).toHaveBeenCalledWith(0, 'Filter 1');
        expect(wrapper.state().active).toBe(false);
    });

    it('should render an add-button with expanded list of available filters', () => {
        const component = (
            <BtnAdd
                available={ [{
                    label: 'Filter 1'
                }, {
                    label: 'Filter 2'
                }] }
                shown
            />
        );
        const wrapper = shallow(component);
        wrapper.setState({ active: true });
        expect(wrapper).toMatchSnapshot();

        document.body.innerHTML = `
            <div>
                body
            </div>
        `;
        const mountedComponent = mount(component);

        document.body.click();
        expect(mountedComponent.state().active).toBe(false);

        document.removeEventListener = jest.fn();
        mountedComponent.unmount();
        expect(document.removeEventListener).toHaveBeenCalledTimes(1);
    });

    it('should render a disabled add-button', () => {
        const wrapper = shallow(
            <BtnAdd
                available={ [{
                    label: 'Filter 1'
                }, {
                    label: 'Filter 2'
                }] }
                disabled
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a left-positioned add-button', () => {
        const wrapper = shallow(
            <BtnAdd
                available={ [{
                    label: 'Filter 1'
                }, {
                    label: 'Filter 2'
                }] }
                position="left"
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a labeled add-button', () => {
        const wrapper = shallow(
            <BtnAdd
                available={ [{
                    label: 'Filter 1'
                }, {
                    label: 'Filter 2'
                }] }
                label="Add more groups"
            />
        );

        expect(wrapper).toMatchSnapshot();
    });
});
