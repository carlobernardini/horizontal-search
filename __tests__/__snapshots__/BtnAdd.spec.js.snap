// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BtnAdd /> should render a default add-button 1`] = `
<div
  className="orizzonte__btn-wrapper"
>
  <button
    className="orizzonte__btn orizzonte__btn-add"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
     
  </button>
</div>
`;

exports[`<BtnAdd /> should render a disabled add-button 1`] = `
<div
  className="orizzonte__btn-wrapper"
>
  <button
    className="orizzonte__btn orizzonte__btn-add orizzonte__btn--disabled"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
     
  </button>
</div>
`;

exports[`<BtnAdd /> should render a labeled add-button 1`] = `
<div
  className="orizzonte__btn-wrapper"
>
  <button
    className="orizzonte__btn orizzonte__btn-add orizzonte__btn--labeled"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Add more groups
  </button>
</div>
`;

exports[`<BtnAdd /> should render a left-positioned add-button 1`] = `
<div
  className="orizzonte__btn-wrapper orizzonte__btn-wrapper--left"
>
  <button
    className="orizzonte__btn orizzonte__btn-add"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
     
  </button>
</div>
`;

exports[`<BtnAdd /> should render an add-button that is (always) shown 1`] = `
<div
  className="orizzonte__btn-wrapper"
>
  <button
    className="orizzonte__btn orizzonte__btn-add orizzonte__btn--shown"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
     
  </button>
</div>
`;

exports[`<BtnAdd /> should render an add-button with expanded list of available filters 1`] = `
<div
  className="orizzonte__btn-wrapper"
>
  <button
    className="orizzonte__btn orizzonte__btn-add orizzonte__btn--shown"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
     
  </button>
  <OrizzonteList
    cache={Object {}}
    doneBtn={true}
    doneBtnLabel={null}
    isFilterGroup={false}
    minWidth={null}
    onApply={[Function]}
    onRemove={[Function]}
    onUpdate={[Function]}
    orientation="right"
    removeBtn={false}
    removeBtnLabel={null}
    syncCacheToGroup={[Function]}
    values={Object {}}
  >
    <a
      className="orizzonte__item-clickable"
      href="#"
      key="orizzonte-add-0"
      onClick={[Function]}
    >
      Filter 1
    </a>
    <a
      className="orizzonte__item-clickable"
      href="#"
      key="orizzonte-add-1"
      onClick={[Function]}
    >
      Filter 2
    </a>
  </OrizzonteList>
</div>
`;
